### PR TITLE
Update to TypeScript 3.0.1

### PR DESCRIPTION
**Pull request by MacChaeger**
I... may have gotten carried away in a few places.

### CODE REVIEW

**Comment by MacChaeger** [comment]
😕 The build passes when I test locally. It has to, because some new feature is making me run `npm test` before every push. Am I out of sync somehow? I even made sure to run `npm install` again.

**Comment by MacChaeger** [comment]
Yeah, I realized that now. I’ll be sure to mention that in the ts-ignore comments so nobody does this in the future

File: chat-formatter.js [hunk 1]
```diff
@@ -87,7 +87,7 @@ class TextFormatter {
 		this.stack = /** @type {FormatSpan[]} */ ([]);
 		this.isTrusted = isTrusted;
 		/** offset of str that's been parsed so far */
-		// @ts-ignore TypeScript bug: can't infer type
+		/** @type {number} */
```
Comment (Zarel) [inline chat-formatter.js]: Is the TypeScript bug still present? We should report it probably...

File: chat-plugins/helptickets.js [hunk 2]
```diff
@@ -41,6 +41,7 @@ let tickets = {};
 let ticketBans = {};
 
 try {
+	/** @type {typeof tickets & typeof ticketBans} */
```
Comment (Zarel) [inline chat-plugins/helptickets.js]: :|

File: chat.js [hunk 3]
```diff
@@ -102,8 +108,11 @@ Chat.multiLinePattern = new PatternTester();
  * Load command files
  *********************************************************/
 
+/** @type {CommandContext} */
```
Comment (Zarel) [inline chat.js]: It really isn't though...

File: data/scripts.js [hunk 4]
```diff
@@ -193,38 +194,43 @@ let BattleScripts = {
...
+					pokemon.addVolatile('followme', pokemon, zPower);
+					break;
+				case 'crit2':
+					pokemon.addVolatile('focusenergy', pokemon, zPower);
+					break;
+				case 'curse':
+					if (pokemon.hasType('Ghost')) {
+						this.heal(pokemon.maxhp, pokemon, pokemon, zPower);
+					} else {
+						this.boost({atk: 1}, pokemon, pokemon, zPower);
```
Comment (Zarel) [inline data/scripts.js]: this is some pretty serious indentation...

File: chat-formatter.js (hunk 1 above)
Comment (MacChaeger) [reply chat-formatter.js]: Yeah, it still doesn't infer that. For now, compromise and leave the type declaration and also a comment saying it's a bug?

File: chat.js (hunk 3 above)
Comment (MacChaeger) [reply chat.js]: Oh, those should be `ChatCommands`, right?

File: chat-formatter.js (hunk 1 above)
Comment (Zarel) [reply chat-formatter.js]: sure

File: chat.js (hunk 3 above)
Comment (Zarel) [reply chat.js]: I think so

File: sim/dex.js [hunk 5]
```diff
@@ -480,6 +480,7 @@ class ModdedDex {
 		}
 		let id = toId(name);
 		let effect;
+		let otherEffects = {reoil: 'Recoil', drain: 'Drain', zpower: 'Z Power'};
```
Comment (Mystifi) [inline sim/dex.js]: Possible typo with "reoil?"

Comment (MacChaeger) [reply sim/dex.js]: Good catch

File: chat-plugins/helptickets.js (hunk 2 above)
Comment (MacChaeger) [reply chat-plugins/helptickets.js]: Ugh, even if I strictly define the property `banned` on `TicketState` as always `undefined`, TypeScript doesn't infer that an object of `TicketState | BannedTicketState` whose `banned` property evaluates to true _must_ be a `BannedTicketState`.

Comment (MacChaeger) [reply chat-plugins/helptickets.js]: Oh, actually, TypeScript _thinks_ that `Object.assign({}, tickets, ticketBans);` returns an object of type `{[k: string]: TicketState} & {[k: string]: BannedTicketState}`, which it thinks is equivalent to `{k: string]: TicketState & BannedTicketState}`, but `Object.assign` never actually creates intersections if the keys are the same; it just overwrites the old value with the new one instead of doing a nested assignment. I'll report that.

Comment (Zarel) [reply chat-plugins/helptickets.js]: > `Object.assign` never actually creates intersections if the keys are the same

That should not be reported; it's intentional. TypeScript doesn't have very powerful semantics for `Object.assign`; you just need to be aware of the caveats.

Comment (Zarel) [reply chat-plugins/helptickets.js]: > TypeScript doesn't infer that an object of `TicketState | BannedTicketState` whose `banned` property evaluates to true _must_ be a `BannedTicketState`.

Yeah, that one is a bug. I couldn't figure out why, but it's why I split them into two tables in the first place, because I couldn't get TypeScript to tell them apart back when they were in one table.

**Comment by sirDonovan** [comment]
@MacChaeger can you rebase this when you get a chance?

**Comment by Zarel** [comment]
This would be really nice to have, but it does need a rebase. :/

**Comment by Zarel** [comment]
Out of curiosity, how did you fix

```
[ts] Type 'Side' is not assignable to type 'Side'. Two different types with this name exist, but they are unrelated.
```

File: users.js [hunk 6]
```diff
@@ -1696,5 +1692,5 @@ let Users = Object.assign(getUser, {
 	}, 1000 * 60 * 30),
 	socketConnect: socketConnect,
 });
-
+// @ts-ignore
```
Comment (Zarel) [inline users.js]: Why does this need to be ts-ignored?

File: sim/pokemon.js [hunk 7]
```diff
@@ -35,7 +35,7 @@ class Pokemon {
 		let pokemonScripts = this.battle.data.Scripts.pokemon;
 		if (pokemonScripts) Object.assign(this, pokemonScripts);
 
-		if (typeof set === 'string') set = {name: set};
+		if (typeof set === 'string') set = new this.battle.Data.PokemonSet({name: set});
```
Comment (Zarel) [inline sim/pokemon.js]: ???

I don't think `PokemonSet` needs to be a class; it's fine as an interface...

File: lib/fs.js [hunk 8]
```diff
@@ -209,7 +209,6 @@ class FSPath {
 	symlinkTo(target) {
 		if (Config.nofswriting) return Promise.resolve();
 		return new Promise((resolve, reject) => {
-			// @ts-ignore TypeScript bug
```
Comment (Zarel) [inline lib/fs.js]: hee, I was actually the one who fixed this bug!

File: dev-tools/globals.ts [hunk 9]
```diff
@@ -538,7 +541,7 @@ interface Move extends Effect, MoveData {
 	zPowered?: boolean
 }
 
-type TemplateAbility = {0: string, 1?: string, H?: string, S?: string}
+type TemplateAbility = {[key: string]: string | undefined, [index: number]: string | undefined, 0: string, 1?: string, H?: string, S?: string}
```
Comment (Zarel) [inline dev-tools/globals.ts]: Ugh I vetoed this sort of thing. Keep the `@ts-ignore`s everywhere; it's better than wrecking this type and making it worse than useless.

File: dev-tools/globals.ts [hunk 10]
```diff
@@ -22,25 +23,27 @@ let TeamValidatorAsync = require('../team-validator-async');
 
 type GenderName = 'M' | 'F' | 'N' | '';
 type StatName = 'hp' | 'atk' | 'def' | 'spa' | 'spd' | 'spe';
-type StatsTable = {hp: number, atk: number, def: number, spa: number, spd: number, spe: number};
-type SparseStatsTable = {hp?: number, atk?: number, def?: number, spa?: number, spd?: number, spe?: number};
-type SparseBoostsTable = {hp?: number, atk?: number, def?: number, spa?: number, spd?: number, spe?: number, accuracy?: number, evasion?: number};
+type StatsTable = {[key:string]: number, hp: number, atk: number, def: number, spa: number, spd: number, spe: number};
+type SparseStatsTable = Partial<StatsTable>;
+type BoostsTable = {[key:string]: number, atk: number, def: number, spa: number, spd: number, spe: number, accuracy: number, evasion: number};
```
Comment (Zarel) [inline dev-tools/globals.ts]: Ditto these two. Just keep the ts-ignores.

File: users.js (hunk 6 above)
Comment (MacChaeger) [reply users.js]: `[ts] Variable 'Users' is used before being assigned.` I started getting this error with the very first beta release of 3.0.0 I used,, and 3.0.1 still has it. I could never manage to fix it properly, so I had to just ignore it.

File: dev-tools/globals.ts (hunk 9 above)
Comment (MacChaeger) [reply dev-tools/globals.ts]: What do you mean by "worse than useless"?

Comment (HoeenCoder) [reply dev-tools/globals.ts]: Its not strict enough to be effective at ensuring that abilities only contain the correct info, see also: https://github.com/Zarel/Pokemon-Showdown/pull/4782#pullrequestreview-143007785

File: sim/pokemon.js (hunk 7 above)
Comment (MacChaeger) [reply sim/pokemon.js]: I needed to construct a `PokemonSet` object in order to properly TS `fastUnpackTeam`, and I figured it would make sense to make a constructor, which let me add a bunch of documentation to all the properties. Looking at it now, I can see a way to accomplish the same thing with a few instances of `//@ts-ignore` instead of making a class, but then we'd lose all that nice documentation I spent at least an hour writing.

File: dev-tools/globals.ts (hunk 9 above)
Comment (MacChaeger) [reply dev-tools/globals.ts]: Oh. I didn't realize that adding a key signature or w/e it's called was a free pass to any and all properties as long as they're numbers. I guess it never occurred to me they'd make such a huge breaking change as to prevent iterating through objects with a strict list of properties, even if all the properties have the same type, and then suggest a change that ruins the point of defining the type in the first place.

File: sim/pokemon.js (hunk 7 above)
Comment (Zarel) [reply sim/pokemon.js]: Can you point me to which line needs `ts-ignore` if PokemonSet were an interface?

Comment (Zarel) [reply sim/pokemon.js]: I'd probably still do `/** @type {PokemonSet} */ ({})` rather than `new PokemonSet`, though.

File: dev-tools/globals.ts (hunk 9 above)
Comment (Zarel) [reply dev-tools/globals.ts]: TypeScript never allowed iterating objects with a strict list of properties, because TypeScript types are interfaces, not two-way strict types.

We've always dealt with that with `ts-ignore`s everywhere.

**Comment by MacChaeger** [comment]
The thing with the "Side is not compatible with Side" error is that it's actually fake. If you remember me talking about "phantom errors" in Dev a while back, that's the one I meant. The error only shows up when you open `battle.js` or make a modification to it, and it goes away as soon as you open or make an edit to any other file. The `npm test` command never catches it, and neither does anything that Travis does. I never actually dealt with it or did anything to stop it from appearing, I just figured out that it's probably meaningless and ignored it.

File: dev-tools/globals.ts (hunk 9 above)
Comment (Zarel) [reply dev-tools/globals.ts]: By which I mean, `{a: 1, b: 'hi'}` satisfies the type `{a: number}`, so iterating through a type with only numbers in no way guarantees that all the properties will be numbers.

File: sim/pokemon.js (hunk 7 above)
Comment (MacChaeger) [reply sim/pokemon.js]: I didn't think that kind of thing worked. Is that a new feature I missed?

File: dev-tools/globals.ts (hunk 10 above)
Comment (MacChaeger) [reply dev-tools/globals.ts]: Is there really not some better way of doing this, though? We're talking _at least_ 100 ignores...

File: dev-tools/globals.ts (hunk 9 above)
Comment (MacChaeger) [reply dev-tools/globals.ts]: Blegh.

Comment (MacChaeger) [reply dev-tools/globals.ts]: By which I mean, thank you for explaining that, it makes sense to me now.

File: dev-tools/globals.ts (hunk 10 above)
Comment (Zarel) [reply dev-tools/globals.ts]: You can cast to `{[key:string]: number}` if you want, in the `for` part of the loop?

Comment (Zarel) [reply dev-tools/globals.ts]: Alternatively: Maybe we should just have a `STAT_NAMES` array somewhere, and then just `for (const stat of STAT_NAMES) {` instead of `for..in`?

File: sim/pokemon.js (hunk 7 above)
Comment (Zarel) [reply sim/pokemon.js]: It's just how you cast in `.js`-TypeScript.

Comment (Zarel) [reply sim/pokemon.js]: It's the equivalent of `<PokemonSet>{}` or `{} as PokemonSet` in real TypeScript.

File: data/abilities.js [hunk 11]
```diff
@@ -3530,8 +3529,7 @@ let BattleAbilities = {
 			if (effect && effect.id === 'toxicspikes') return;
 			if (status.id === 'slp' || status.id === 'frz') return;
 			this.add('-activate', target, 'ability: Synchronize');
-			// @ts-ignore
-			source.trySetStatus(status, target, {status: status.id, id: 'synchronize'});
+			source.trySetStatus(status, target, Object.assign({status: status.id}, this.effect));
```
Comment (Zarel) [inline data/abilities.js]: Please don't make the code slower for TypeScript.

This is why `ts-ignore` exists. :|

(Although I would just cast here)

File: data/moves.js [hunk 12]
```diff
@@ -1294,19 +1295,11 @@ let BattleMovedex = {
 						this.add('-miss', pokemon, target);
 						return false;
 					}
-					let moveData = {
-						id: 'bide',
-						name: "Bide",
-						accuracy: true,
+					Object.assign(move, {
```
Comment (Zarel) [inline data/moves.js]: I'm pretty sure overwriting the `move` object here is an actual bug. Leave it as a copy.

File: data/moves.js [hunk 13]
```diff
@@ -2341,9 +2334,8 @@ let BattleMovedex = {
 			},
 		},
 		isZ: "kommoniumz",
-		secondary: {
-			// Sheer Force negates the selfBoost even though it is not secondary
-		},
+		// @ts-ignore ~ Sheer Force negates the selfBoost even though it is not secondary
+		secondary: {},
```
Comment (Zarel) [inline data/moves.js]: Why does this need a `ts-ignore`? `secondary` should be allowed to be an empty object.

File: data/moves.js [hunk 14]
```diff
@@ -5490,9 +5477,9 @@ let BattleMovedex = {
 			} else {
 				if (!move.secondaries) move.secondaries = [];
 				if (item.fling.status) {
-					move.secondaries.push({status: item.fling.status});
+					move.secondaries.push({status: item.fling.status, chance: 100});
```
Comment (Zarel) [inline data/moves.js]: `chance` should be optional for a secondary...

Comment (Zarel) [reply data/moves.js]: Is this why you needed a `ts-ignore` on `secondary: {}`?

File: data/moves.js [hunk 15]
```diff
@@ -6823,7 +6806,9 @@ let BattleMovedex = {
 			let sourceBoosts = {};
 
 			for (const stat of ['def', 'spd']) {
+				// @ts-ignore
```
Comment (Zarel) [inline data/moves.js]: Okay, this should just be `/* StatName[] */ (['def', 'spd'])`

File: data/moves.js [hunk 16]
```diff
@@ -10294,8 +10278,7 @@ let BattleMovedex = {
 				let move = exports.BattleMovedex[i];
 				if (i !== move.id) continue;
 				if (move.isZ || move.isNonstandard) continue;
-				// @ts-ignore
-				if (effect.noMetronome.includes(move.id)) continue;
+				if (effect.noMetronome && effect.noMetronome.includes(move.id)) continue;
```
Comment (Zarel) [inline data/moves.js]: Also, your checking instead of `ts-ignore` I'm generally against. `ts-ignore` is correct, this should be an assertion, not a check.

File: data/moves.js [hunk 17]
```diff
@@ -10518,17 +10501,16 @@ let BattleMovedex = {
 		effect: {
 			duration: 1,
 			noCopy: true,
-			onStart: function (target, source, source2, move) {
+			onStart: function () {
```
Comment (Zarel) [inline data/moves.js]: Does TS not allow empty args anymore? We should probably allow them for consistency.

File: mods/gennext/moves.js [hunk 18]
```diff
@@ -677,12 +677,12 @@ let BattleMovedex = {
 					}
 					this.add('-end', pokemon, 'Bide');
 					let target = this.effectData.sourceSide.active[this.effectData.sourcePosition];
-					/**@type {Move} */
-					// @ts-ignore
-					let moveData = {
+					Object.assign(move, {
```
Comment (Zarel) [inline mods/gennext/moves.js]: Ditto, you changed a clone into a mutation, and mutation is bad.

File: mods/mixandmega/items.js [hunk 19]
```diff
@@ -42,13 +42,12 @@ let BattleItems = {
 				pokemon.formeChange(template, this.effect, true);
 				pokemon.baseTemplate = template;
 				this.add('-start', pokemon, 'Red Orb', '[silent]');
-				// @ts-ignore
-				let oTemplate = this.getTemplate(pokemon.illusion || pokemon.originalSpecies);
+				let oTemplate = this.getTemplate(pokemon.originalSpecies);
```
Comment (Zarel) [inline mods/mixandmega/items.js]: This seems wrong? What are you trying to achieve here?

File: mods/mixandmega/scripts.js [hunk 20]
```diff
@@ -38,8 +38,8 @@ let BattleScripts = {
 		}
 
 		// Do we have a proper sprite for it?
-		// @ts-ignore
-		if (this.getTemplate(pokemon.canMegaEvo).baseSpecies === pokemon.originalSpecies || isUltraBurst) {
+		// @ts-ignore TypeScript bug: TS doesn't infer that if isUltraBurst is false, pokemon.canMegaEvo must be a string
```
Comment (Zarel) [inline mods/mixandmega/scripts.js]: I wouldn't call this a TS bug; TS was never designed to handle these sorts of complex relationships. This is just an assertion.

File: mods/mixandmega/scripts.js [hunk 21]
```diff
@@ -99,12 +99,9 @@ let BattleScripts = {
 		} else if (deltas.type) {
 			template.types = [template.types[0], deltas.type];
 		}
-		let baseStats = template.baseStats;
-		// @ts-ignore
-		template.baseStats = {};
-		for (let statName in baseStats) {
+		for (let statName in template.baseStats) {
```
Comment (Zarel) [inline mods/mixandmega/scripts.js]: You just transformed a copy into a mutation. :| There's a reason it was a copy.

**Comment by MacChaeger** [comment]
Ok, I don’t care about this any more. Do whatever.

**Comment by Zarel** [comment]
Did I say the wrong thing? I can take this over, but for the record, I do appreciate everything you've done here.

**Comment by KrisXV** [comment]
I can do it too if you’d like

File: chat.js [hunk 22]
```diff
@@ -784,7 +792,11 @@ class CommandContext {
 	 * @param {User?} [targetUser]
 	 */
 	canTalk(message = null, room = null, targetUser = null) {
-		// @ts-ignore
+		if (this.room instanceof Rooms.GlobalRoom) {
```
Comment (Zarel) [inline chat.js]: This is also not equivalent to the previous code, because `this.room` _can_ be a global room as long as `room` isn't.

File: data/scripts.js [hunk 23]
```diff
@@ -841,7 +804,7 @@ let BattleScripts = {
 	},
 
 	calcRecoilDamage: function (damageDealt, move) {
-		// @ts-ignore
+		if (!move.recoil) throw new TypeError(`Tried to calculate recoil damage, but move ${move.name} had no recoil data.`);
```
Comment (Zarel) [inline data/scripts.js]: This does the same thing as `@ts-ignore` but has worse performance.

File: lib/streams.js [hunk 24]
```diff
@@ -411,10 +414,6 @@ class ReadWriteStream extends ReadStream {
 	}
 }
 
-// TypeScript bug: no generic support
```
Comment (Zarel) [inline lib/streams.js]: Hey, this is still a bug!

File: mods/gen4/items.js [hunk 25]
```diff
@@ -48,8 +48,7 @@ let BattleItems = {
 					priority: action.priority + 0.1,
 					pokemon: action.pokemon,
 					move: action.move,
-					// @ts-ignore
-					target: action.target,
+					target: action.targetLoc,
```
Comment (Zarel) [inline mods/gen4/items.js]: This looks like a mechanics bugfix... should probably not be bundled into a bunch of unrelated refactors.

Comment (Zarel) [reply mods/gen4/items.js]: In hindsight you might as well, but the correct fix is `targetLoc: action.targetLoc`.

File: punishments.js [hunk 26]
```diff
@@ -1486,11 +1483,12 @@ Punishments.isRoomBanned = function (user, roomid) {
  * options.publicOnly will make this only return public room punishments.
  * options.checkIps will also check the IP of the user for IP-based punishments.
  *
- * @param {User | string} user
+ * @param {User? | string} user
```
Comment (Zarel) [inline punishments.js]: It's a bad idea to change a call signature in a way that would obviously be a bug if actually used that way.

Just make a temporary variable...

File: punishments.js [hunk 27]
```diff
@@ -1486,11 +1483,12 @@ Punishments.isRoomBanned = function (user, roomid) {
...
  * @param {Object?} options
  * @return {Array}
  */
 Punishments.getRoomPunishments = function (user, options) {
+	if (typeof user === 'string') user = Users(user);
```
Comment (Zarel) [inline punishments.js]: Also this is a bug, we intentionally don't do this because `getRoomPunishments` is supposed to be able to find punishments for offline users.

File: rooms.js [hunk 28]
```diff
@@ -865,8 +867,7 @@ class GlobalRoom extends BasicRoom {
 	startLockdown(err, slow = false) {
 		if (this.lockdown && err) return;
 		let devRoom = Rooms('development');
-		// @ts-ignore
-		const stack = (err ? Chat.escapeHTML(err.stack).split(`\n`).slice(0, 2).join(`<br />`) : ``);
+		const stack = (err.stack ? Chat.escapeHTML(err.stack).split(`\n`).slice(0, 2).join(`<br />`) : ``);
```
Comment (Zarel) [inline rooms.js]: This will crash if `err` is `null`, which, contrary to the function signature, it might be.

Comment (Zarel) [reply rooms.js]: (Specifically, it will be if you use `/lockdown`.)

File: sim/pokemon.js [hunk 29]
```diff
@@ -166,10 +169,11 @@ class Pokemon {
 		this.speed = 0;
 		this.abilityOrder = 0;
 
-		set.level = this.battle.clampIntRange(set.forcedLevel || set.level || 100, 1, 9999);
+		set.level = this.battle.clampIntRange(set.level || 100, 1, 9999);
```
Comment (Zarel) [inline sim/pokemon.js]: ???

File: test/application/team-validator.js [hunk 30]
```diff
@@ -48,15 +48,15 @@ describe('Team Validator', function () {
 	});
 
 	it('should validate Gen 2 IVs', function () {
-		let team = Dex.fastUnpackTeam('|raikou|||hiddenpowerwater||||14,28,26,,,|||');
+		let team = Dex.fastUnpackTeam('|raikou|||hiddenpowerwater||1||14,28,26,,,|||');
```
Comment (Zarel) [inline test/application/team-validator.js]: A refactor should not make valid teams invalid. Backwards compatibility is important!

**Comment by Zarel** [comment]
Okay, I ripped out around half of the logic changes because they led to bugs. I also made a simplification of the data model. Everything else is _probably_ good.

File: data/abilities.js [hunk 31]
```diff
@@ -3529,7 +3529,7 @@ let BattleAbilities = {
 			if (effect && effect.id === 'toxicspikes') return;
 			if (status.id === 'slp' || status.id === 'frz') return;
 			this.add('-activate', target, 'ability: Synchronize');
-			source.trySetStatus(status, target, Object.assign({status: status.id}, this.effect));
+			source.trySetStatus(status, target);
```
Comment (urkerab) [inline data/abilities.js]: Synchronize is a primary effect so it needs to pass an object with the appropriate status to `trySetStatus` otherwise it's processed as if it's a secondary effect.

Comment (Zarel) [reply data/abilities.js]: @urkerab Can you explain the difference between "primary effect" and "secondary effect"? That's not a difference I remember implementing when I made the battle engine.
